File: dawn_go/ukkoken.go
package main

import "fmt"

// On-line construction of suffix trees, Ukkoken '95

type node struct {
    start      int
    len        int
    children   map[rune]*node
    suffixLink *node
    parent     *node
}

func createNode(start, len int, parent *node) *node {
    return &node{
        start:      start,
        len:        len,
        children:   make(map[rune]*node),
        suffixLink: nil,
        parent:     parent,
    }
}

func splitNode(n *node, start, len int, nodeLabel rune) *node {
    newNode := createNode(start, len, n)
    remainingLen := n.len - len - 1
    n.len = len
    child := createNode(n.start+len+1, remainingLen, newNode)
    n.children[nodeLabel] = child
    newNode.suffixLink = n.suffixLink
    n.suffixLink = newNode
    return newNode
}

func findNode(root *node, start, len int, s string) *node {
    if len == 0 {
        return root
    }
    child, ok := root.children[rune(s[start])]
    if !ok {
        return nil
    }
    childLen := child.len
    if childLen >= len {
        return child
    }
    return findNode(child, child.start+childLen+1, len-childLen-1, s)
}

func followLinks(n *node) *node {
    if n.suffixLink == nil {
        return n
    }
    return followLinks(n.suffixLink)
}

func updateTree(s string, i int, root *node) {
    len := len(s)
    n := followLinks(root)
    remaining := len - i
    if remaining > 0 {
        leaf := createNode(i, remaining, n)
        nodeLabel := rune(s[i])
        child, ok := n.children[nodeLabel]
        if ok {
            start := child.start
            len := child.len
            foundNode := findNode(root, start, len, s)
            if foundNode != nil && s[start:start+len] == s[i:i+len] {
                updateTree(s, i+len+1, child)
            } else if foundNode != nil {
                newNode := splitNode(foundNode, start, len, nodeLabel)
                leaf = createNode(i, remaining, newNode)
                newNode.children[nodeLabel] = leaf
            } else {
                n.children[nodeLabel] = leaf
            }
        } else {
            n.children[nodeLabel] = leaf
        }
    }
}

func buildSuffixTree(s string) *node {
    root := createNode(0, 0, nil)
    for i := range s {
        updateTree(s, i, root)
    }
    return root
}

File: dawn_python/ukkoken.py

# On-line construction of suffix trees, Ukkoken '95

class Node:
    def __init__(self, start, length, parent=None):
        self.start = start
        self.length = length
        self.children = {}
        self.suffix_link = None
        self.parent = parent

def create_node(start, length, parent):
    return Node(start, length, parent)

def split_node(node, start, length, node_label):
    new_node = create_node(start, length, node.parent)
    remaining_length = node.length - length - 1
    node.length = length
    child = create_node(node.start + length + 1, remaining_length, new_node)
    node.children[node_label] = child
    new_node.suffix_link = node.suffix_link
    node.suffix_link = new_node
    return new_node

def find_node(root, start, length, s):
    if length == 0:
        return root
    child = root.children.get(s[start])
    if child:
        child_length = child.length
        if child_length >= length:
            return child
        return find_node(child, child.start + child_length + 1, length - child_length - 1, s)
    return None

def follow_links(node):
    if node.suffix_link:
        return follow_links(node.suffix_link)
    return node

def update_tree(s, i, root):
    length = len(s)
    node = follow_links(root)
    remaining = length - i
    if remaining > 0:
        leaf = create_node(i, remaining, node)
        node_label = s[i]
        child = node.children.get(node_label)
        if child:
            start = child.start
            length = child.length
            found_node = find_node(root, start, length, s)
            if found_node and s[start:start + length] == s[i:i + length]:
                update_tree(s, i + length + 1, child)
            elif found_node:
                new_node = split_node(found_node, start, length, node_label)
                leaf = create_node(i, remaining, new_node)
                new_node.children[node_label] = leaf
        else:
            node.children[node_label] = leaf

def build_suffix_tree(s):
    root = create_node(0, 0, None)
    for i in range(len(s)):
        update_tree(s, i, root)
    return root